### ISSUE TITLE
GnGeo compatible with MAME?

### ISSUE THREAD

**Issue by videlanicolas** [dciabrin/ngdevkit#110]
Hi there! I compiled `ngdevkit` and downloaded and compiled the examples. They work with GnGeo.

When I move the built 01-helloworld ROM (everything under `rom/`) to MAME under `/usr/share/games/mame/roms/puzzledp` and try to emulate it I get the following error:

```bash
$ mame -debug -nofilter -window puzzledp
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
Debug Build: Disabling input grab for -debug
Debug Build: Disabling input grab for -debug
Debug Build: Disabling input grab for -debug
202-p1.p1 WRONG CHECKSUMS:
    EXPECTED: CRC(2b61415b) SHA1(0e3e4faf2fd6e63407425e1ac788003e75aeeb4f)
       FOUND: CRC(6cc4a5bc) SHA1(d125070aca7506e06da45c0263ee19cbf19319ca)
202-s1.s1 WRONG CHECKSUMS:
    EXPECTED: CRC(cd19264f) SHA1(531be2305cd56d332fb7a53ab924214ade34a9e8)
       FOUND: CRC(2e4238d1) SHA1(f556cd9afcf53c3a1128cdaa639d49fb3abfa3ed)
sfix.sfix NOT FOUND (tried in puzzledp neogeo)
000-lo.lo WRONG LENGTH (expected: 00020000 found: 00010000)
000-lo.lo WRONG CHECKSUMS:
    EXPECTED: CRC(5a86cff2) SHA1(5992277debadeb64d1c1c64b0a92d9293eaf7e4a)
       FOUND: CRC(e09e253c) SHA1(2b1c719531dac9bb503f22644e6e4236b91e7cfc)
sm1.sm1 WRONG CHECKSUMS:
    EXPECTED: CRC(94416d67) SHA1(42f9d7ddd6c0931fd64226a60dc73602b2819dcf)
       FOUND: CRC(97cf998b) SHA1(977387a7c76ef9b21d0b01fa69830e949a9a9626)
202-m1.m1 WRONG CHECKSUMS:
    EXPECTED: CRC(9c0291ea) SHA1(3fa67c62acba79be6b3a98cc1601e45569fa11ae)
       FOUND: CRC(0af237d9) SHA1(4b17eacd9a0de2a4ae2dcee8f21c7ef7528ecbcc)
202-v1.v1 WRONG CHECKSUMS:
    EXPECTED: CRC(debeb8fb) SHA1(49a3d3578c087f1a0050168571ef8d1b08c5dc05)
       FOUND: CRC(75660aac) SHA1(6a521e1d2a632c26e53b83d2cc4b0edecfc1e68c)
202-c1.c1 WRONG CHECKSUMS:
    EXPECTED: CRC(cc0095ef) SHA1(3d86f455e6db10a2449b775dc386f1826ba3b62e)
       FOUND: CRC(6ffb6b00) SHA1(a98db4387928b66f9f0ba74846e0f07065ec8a29)
202-c2.c2 WRONG CHECKSUMS:
    EXPECTED: CRC(42371307) SHA1(df794f989e2883634bf7ffeea48d6bc3854529af)
       FOUND: CRC(5cf57ada) SHA1(333a7933caf110d0421a5c7c0b8bf6e07108f96d)
Fatal error: Required files are missing, the machine cannot be run.
```

Running other NeoGeo games on MAME works.

Are these examples supposed to work on MAME or only on GnGeo?

**Comment by dciabrin** [comment]
TL;DR: Yes the examples are compatible with both GnGeo and MAME.
you can run the example from the ./01-helloworld directory directly with e.g.:
```
mame -rp rom puzzledp -w -noautosave -skip_gameinfo
```


The checksum complains are normal, both GnGeo and MAME have a fixed list of games that you're allowed to emulate, so the example ROMs reuse names, and obviously their checksum don't match.

What's odd though is the neogeo BIOS that is being loaded by MAME doesn't seem to be the one built by ngdevkit.
```
000-lo.lo WRONG LENGTH (expected: 00020000 found: 00010000)
```
MAME has loaded a zip that contains a zoom precalc with is 64KB long, whereas ngdevkit's stub BIOS is 128KB as expected.
```
$ unzip -l rom/neogeo.zip
Archive:  rom/neogeo.zip
nullbios 0.4, an open source BIOS for Neo Geo
https://github.com/dciabrin/ngdevkit
  Length      Date    Time    Name
---------  ---------- -----   ----
   131072  03-18-2024 18:26   000-lo.lo
   131072  03-18-2024 18:26   sfix.sfix
   131072  03-18-2024 18:26   sm1.sm1
   131072  03-18-2024 18:26   sp-s2.sp1
   131072  03-18-2024 18:26   aes-bios.bin
---------                     -------
   655360                     5 files
```

**Comment by videlanicolas** [comment]
> TL;DR: Yes the examples are compatible with both GnGeo and MAME. you can run the example from the ./01-helloworld directory directly with e.g.:
> 
> ```
> mame -rp rom puzzledp -w -noautosave -skip_gameinfo
> ```

It works! Thanks!

> MAME has loaded a zip that contains a zoom precalc with is 64KB long, whereas ngdevkit's stub BIOS is 128KB as expected.

Searching through Google I found similar issues: https://www.google.com/search?q=%22000-lo.lo+WRONG+LENGTH%22

Seems like an old version of the BIOS?

> nullbios 0.4, an open source BIOS for Neo Geo

So this BIOS is supposed to replace the original NeoGeo BIOS? But given it's open source it has no license implications for its use?

**Comment by dciabrin** [comment]
> > MAME has loaded a zip that contains a zoom precalc with is 64KB long, whereas ngdevkit's stub BIOS is 128KB as expected.
> 
> Searching through Google I found similar issues: https://www.google.com/search?q=%22000-lo.lo+WRONG+LENGTH%22
> 
> Seems like an old version of the BIOS?
> 

Most likely, yes

> > nullbios 0.4, an open source BIOS for Neo Geo
> 
> So this BIOS is supposed to replace the original NeoGeo BIOS? But given it's open source it has no license implications for its use?

This is a made-from-scratch, very minimal stub that is sufficient to make emulators run your programs without requiring the original BIOS. It provides a zoom lookup table for the emulators, an empty sound driver,  and it implements a handful of functions [1] e.g. read inputs and reset the watchdog timer for you. It does not implement the entire API from the original BIOS like memcard load/save, multiplayer inputs, and whatnot.

[1] https://wiki.neogeodev.org/index.php?title=BIOS_calls

**Comment by videlanicolas** [comment]
Would this open source BIOS work for MAME or only for GnGeo? I just tried replacing it in MAME and it's throwing errors, as if MAME can't find some necessary files to run the game.

The reason is that I like the MAME debug console, although I'm fine using gdb if I can't run the game on MAME with this BIOS.

I'm looking to develop a game for the MVS, and honestly don't care about storing saves or highscores, so I'm fine not having memcard calls.

**Comment by dciabrin** [comment]
This is supposed to work for recent MAME (I use 0.263) and the GnGeo fork that ships with ngdevkit. I'm using this stub BIOS with MAME's z80 debugger to develop nullsound, the sound driver that comes with the devkit.
The BIOS should contain the necessary file names to boot the emulator in AES or MVS mode. I only tested AES on GnGeo, but both emulators are similar, I expect it to work fine on MAME.

However, you can't just use this BIOS in place of the original and expect existing games/ROMs to work. This won't work.
Makes sense, or are you hitting a specific issue?

**Comment by videlanicolas** [comment]
> Makes sense, or are you hitting a specific issue?

This makes sense, I was hoping it worked until the game made a call to an unknown BIOS call and stop the emulation.

I removed all files from `roms/neogeo` and replaced it with the files inside `neogeo.zip` that came from the `01-helloworld` example:

```bash
$ /usr/share/games/mame/roms/neogeo$ unzip neogeo.zip 
Archive:  neogeo.zip
nullbios 0.4, an open source BIOS for Neo Geo
https://github.com/dciabrin/ngdevkit
  inflating: 000-lo.lo               
  inflating: aes-bios.bin            
  inflating: sfix.sfix               
  inflating: sm1.sm1                 
  inflating: sp-s2.sp1
```

Then copied the helloworld example rom contents under `roms/puzzledp` (I thought the folder name had to match the game name, but I don't think this is necessary):

```bash
$ /usr/share/games/mame/roms/puzzledp$ cp ~/examples/01-helloworld/rom/* .
$ /usr/share/games/mame/roms/puzzledp$ ls
202-c1.bin  202-c1.c1  202-c2.bin  202-c2.c2  202-m1.bin  202-m1.m1  202-p1.bin  202-p1.p1  202-s1.bin  202-s1.s1  202-v1.bin  202-v1.v1  neogeo.zip  puzzledp.zip
```

Then run the game with mame:

```bash
$ mame -version
0.242 (unknown)
$ mame -rp rom puzzledp -w -noautosave -skip_gameinfo -debug
ALSA lib confmisc.c:855:(parse_card) cannot find card '0'
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_card_inum returned error: No such file or directory
ALSA lib confmisc.c:422:(snd_func_concat) error evaluating strings
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1334:(snd_func_refer) error evaluating name
ALSA lib conf.c:5178:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5701:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2664:(snd_pcm_open_noupdate) Unknown PCM default
Debug Build: Disabling input grab for -debug
Debug Build: Disabling input grab for -debug
Debug Build: Disabling input grab for -debug
202-p1.p1 NOT FOUND (tried in puzzledp neogeo)
202-s1.s1 NOT FOUND (tried in puzzledp neogeo)
sfix.sfix NOT FOUND (tried in puzzledp neogeo)
000-lo.lo NOT FOUND (tried in puzzledp neogeo)
sp-s2.sp1 NOT FOUND (tried in puzzledp neogeo)
sm1.sm1 NOT FOUND (tried in puzzledp neogeo)
202-m1.m1 NOT FOUND (tried in puzzledp neogeo)
202-v1.v1 NOT FOUND (tried in puzzledp neogeo)
202-c1.c1 NOT FOUND (tried in puzzledp neogeo)
202-c2.c2 NOT FOUND (tried in puzzledp neogeo)
Fatal error: Required files are missing, the machine cannot be run.
```

I'll keep debugging this when I have more time, but you answered my original question here (thanks!) so I think we can close this bug.
